model/entities/balance.py

```diff
@@ -43,6 +46,17 @@ def __combine__(self, other: "Balance", combinator: Callable[[int, int], int]):
             for currency in set(self.keys()).union(other.keys())
         })
 
+    def values_in_usd(self, market_prices: Dict[Pair, float]):
```

Comment:
I kinda dislike the two code paths here. I would try to think about this a bit more general. First I'd try to create an object / function that gives me the exchange rate between any two `Currency`s (either directly or implicitly).
So here I wouldn't actually look into the market_prices directly. I would actually pass in a callable, something like:

```python
# in utils/exchange_rate
Exchanger = Callable[[Currency, Currency], float]

def build_exchanger(market_prices: Dict[Pair, float]) -> Exchanger:
  def get_exchange_rate(from: Currency, to: Currency) -> float:
    if market_prices.has(Pair(from, to)) or market_prices.has(Pair(to, from)):
      # direct pair
      return market_price.get(Pair(from, to)) or (1 / market_prices.get(Pair(to, from)))
    else:
       # implicit pair
       for bridge_asset in ...:
         if market_price.has(Pair(from, bridge_asset)) and market_price.has(Pair(to, bridge_asset)):
            ...
     assert False, f"Can not determine rate from {from} to {to}" # if it doesn't return above it will fail
   return get_exchange_rate

   
# in balance:
class Balance:
  def convert_to(self, target_currency: Currency, get_exchange_rate: Exchanger) -> Dict[Currency, float]:
    return {
      key: inventory * get_exchange_rate(key, target_currency)
      for key, inventory in self.items()
    }


# when calling
some_balance.convert_to(Fiat.USD, build_exchanger(market_prices))
```

I feel like something like this would incapsulate concerns a bit better.

Reply:
To be honest, idealy I would like to extend `Pair` to be able to perform correct Pair operations:

```python
class Pair():
    base: Currency
    quote: Currency
    rate: float

    def __init__(self, base, quote, rate):
        self.base = base
        self.quote = quote
        self.rate = rate

    def __str__(self):
        return f"{self.base.value}_{self.quote.value}"

    def ___mul__(self, second_pair):
        if self.quote == second_pair.base:
            base = self.base
            quote = second_pair.quote
            rate = self.rate * second_pair.rate
        elif self.quote == second_pair.quote:
            

        return Pair(base, quote, rate)
```

Reply:
If we go all general, I prefer this approach. What do you think?
I tried to implement only what we need right now which is the USD value of of the reserve by keeping the policies tidy

Reply:
Hmm I think Pair should be a base type but maybe you'd have something like:
```
class Rate():
  paid: Pair,
  value: float
   ...
```